### PR TITLE
Upgrade to CodeMirror 6 release

### LINKED ISSUE #612: Tooltips for diagnostics linger after they've been fixed
Tooltips for diagnostics linger after they've been fixed which missed the opportunity to clearly signal that the issue has been fixed.

They also don't update if you move from one error state to another. I'm less certain that's essential but worth some thought.

The red underlines do update as does the diagnostic in CM's diagnostics panel if that's open, so this is a tooltip handling issue.

Spotted reviewing https://github.com/microbit-foundation/python-editor-next/pull/563 but not new there.

### LINKED ISSUE #697: Upgrade CodeMirror
There are breaking changes that will affect us in various trivial ways:
https://discuss.codemirror.net/t/release-0-20-0/4302

I think we should hold off on this for a few weeks, not least because it might affect the Oxford student team building on the editor, but we should do it in time to fix any issues surfaced by V3 live release.

### LINKED ISSUE #716: Gutter tooltips shouldn't have a hover delay
Gutter tooltips shouldn't have a hover delay as there's no reason to expect some other interaction.

### LINKED ISSUE #753: Editor size unexpectedly short of content
1. Start with a blank narrow code area (reload the page if required so it's the initial size)
2. Type nested "if True" until the editor scrolls
3. Note the current line outline end is to the left of the cursor position

If code structure highlighting is enabled then it breaks at this point.

If you check the DOM then the content editable is unexpectedly narrower than the content that made it scroll.

Split out from https://github.com/microbit-foundation/python-editor-next/issues/253

### LINKED ISSUE #995: Code structure highlight stops prematurely in narrow window
## Bug Description

If the browser window width is small enough (or zoomed in) and code lines are long enough to require horizontal scrolling, then sometimes the code structure highlight stops before the end of the line. 

image

video example showing highlight state:
https://www.loom.com/share/0227af771b3e446db85189a665068187

## How To Reproduce

I can't repro this just by pasting in an example that previously caused the issue, it seems to depend on some previous state. This code has reproduced the bug for me if I type it out line by line (starting with the sample code and adding to the bottom of the editor) with browser zoom set at 175% in Chrome. I also tried it with a slightly different window size/zoom and it didn't repro that time - I have not been able to pin down the exact requirements.

``` python
# Imports go at the top
from microbit import *


# Code in a 'while True:' loop repeats forever
while True:
    display.show(Image.HEART)
    sleep(1000)
    display.scroll('Hello')
    if False:
        print("no")
    if True:
        if True:
            sleep(100)
            print("hello my name is blasduioajdioas djiosa djisoa dj ioasd jio asd")
        
```

### PR DESCRIPTION
**Pull request by microbit-robert**
Closes #995.
Closes #753.
Closes #697.
Closes #612. 
Closes #716.

### CODE REVIEW

File: src/editor/codemirror/lint/lint.ts [hunk 1]
```diff
@@ -1,11 +1,12 @@
 /**
-Modified copy of the lint extension from CodeMirror 0.19.6
-git revision 084c4ad59be7562c739d2e8fff12c7852e106b15
-sha1 of source lint.ts facbdc305d2b5a318dbcb9bcd41f992073ecfff5
+Modified copy of the lint extension from CodeMirror 6.0.0
+git revision 15319f1202f9fa6654e8d1328a0c506ed3d04e7f
+sha1 of source lint.ts facbdc305d2b5a318dbcb9bcd41f992073ecfff5 NEEDS UPDATING!!
```
Comment (microbit-robert) [inline src/editor/codemirror/lint/lint.ts:4]: Needs updating.

File: src/editor/codemirror/lint/lint.ts [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
 
 To aid comparison to the original, this file is not auto-formatted
 and we've disabled our eslint rules.
 
+// REVIEW THESE
 Changes:
```
Comment (microbit-robert) [inline src/editor/codemirror/lint/lint.ts:10]: Changes need careful review and validation.

File: src/editor/codemirror/lint/lint.ts [hunk 3]
```diff
@@ -67,10 +66,13 @@ export interface Diagnostic {
...
   /// The message associated with this diagnostic.
   message: string
+  /// An optional custom rendering function that displays the message
+  /// as a DOM node.
+  renderMessage?: () => Node
   /// An optional array of actions that can be taken on this
   /// diagnostic.
   actions?: readonly Action[]
-  /// Tags control alternative presentations for diagnostics.
+    /// Tags control alternative presentations for diagnostics.
```
Comment (microbit-robert) [inline src/editor/codemirror/lint/lint.ts:75]: Formatting fix required.

File: src/editor/codemirror/lint/lint.ts [hunk 4]
```diff
@@ -606,9 +647,9 @@ const baseTheme = EditorView.baseTheme({
   ".cm-lintRange-error": { backgroundImage: underline("#d11") },
   ".cm-lintRange-warning": { backgroundImage: underline("orange") },
   ".cm-lintRange-info": { backgroundImage: underline("#999") },
-  // Nothing for .cm-lintRange-hint, they're styled based on tags only.
-  ".cm-lintRange-unnecessary": { opacity: 0.5 },
-  ".cm-lintRange-deprecated": { textDecoration: "line-through" },
+    // Nothing for .cm-lintRange-hint, they're styled based on tags only.
+    ".cm-lintRange-unnecessary": { opacity: 0.5 },
+    ".cm-lintRange-deprecated": { textDecoration: "line-through" },
```
Comment (microbit-robert) [inline src/editor/codemirror/lint/lint.ts:652]: Formatting.

File: src/editor/codemirror/lint/lint.ts [hunk 5]
```diff
@@ -857,13 +909,15 @@ const lintGutterConfig = Facet.define<LintGutterConfig, Required<LintGutterConfi
...
   }
-});
+})
 
 /// Returns an extension that installs a gutter showing markers for
 /// each line that has diagnostics, which can be hovered over to see
 /// the diagnostics.
 export function lintGutter(config: LintGutterConfig = {}): Extension {
-  return [lintGutterConfig.of(config), lintGutterMarkers, lintGutterExtension, lintGutterTheme, lintGutterTooltip, editingLineState, editingLinePlugin]
+  return [lintGutterConfig.of(config), lintGutterMarkers, lintGutterExtension, lintGutterTheme, lintGutterTooltip, editingLineState, editingLinePlugin,]
```
Comment (microbit-robert) [inline src/editor/codemirror/lint/lint.ts:922]: Remove trailing comma.

File: src/editor/codemirror/lint/lint.ts (hunk 3 above)
Comment (microbit-robert) [reply src/editor/codemirror/lint/lint.ts]: [`1bca322`](https://github.com/microbit-foundation/python-editor-v3/pull/1016/commits/1bca32266d2cafaa02e24a2990cf0ed87cb64728)

File: src/editor/codemirror/lint/lint.ts (hunk 2 above)
Comment (microbit-robert) [reply src/editor/codemirror/lint/lint.ts]: Changes still apply and appear to work as expected.

**Comment by microbit-robert** [comment]
Leaving as a draft to avoid any accidental merge until we're happy we've tested this enough.

File: package.json [hunk 6]
```diff
@@ -7,19 +7,12 @@
...
-    "@codemirror/highlight": "^0.19.7",
-    "@codemirror/history": "^0.19.2",
-    "@codemirror/lang-python": "^0.19.4",
-    "@codemirror/language": "^0.19.7",
-    "@codemirror/panel": "^0.19.1",
-    "@codemirror/state": "^0.19.9",
-    "@codemirror/view": "^0.19.44",
+    "@codemirror/autocomplete": "^6.3.0",
+    "@codemirror/commands": "6.1.1",
+    "@codemirror/lang-python": "^6.0.2",
```
Comment (microbit-matt-hillsdon) [inline package.json:12]: We should review the diff for lang-python and see if any grammar changes might affect us, e.g. code structure highlighting, actual highlighting (which relies on tags in the grammar), indentation.

Comment (microbit-matt-hillsdon) [reply package.json:12]: Reviewed, nothing significant, though the above are all good areas to test.

**Comment by microbit-matt-hillsdon** [comment]
There's a subtle code structure highlighting regression that I think is worth looking at:

New:
image

Old:
image

**Pull request merged**